### PR TITLE
updated aside.go

### PR DESCRIPTION
**Pull request by LegendHawk27**
Added an option to ClientOption to use Lua scripts for acquiring locks, enhancing compatibility with older Redis versions that do not support the SET NX GET command.

### CODE REVIEW

File: rueidisaside/aside.go [hunk 1]
```diff
@@ -114,7 +117,12 @@ func (c *Client) keepalive() (id string, err error) {
 	c.mu.Unlock()
 	if id == "" {
 		id = PlaceholderPrefix + ulid.Make().String()
-		if err = c.client.Do(c.ctx, c.client.B().Set().Key(id).Value("").Px(c.ttl).Build()).Error(); err == nil {
```
Comment (rueian) [inline rueidisaside/aside.go:117]: This one doesn't need to be replaced by the lua script.

File: rueidisaside/aside.go [hunk 2]
```diff
@@ -197,6 +209,7 @@ func (c *Client) Close() {
 const PlaceholderPrefix = "rueidisid:"
 
 var (
-	delkey = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("DEL",KEYS[1]) else return 0 end`)
-	setkey = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("SET",KEYS[1],ARGV[2],"PX",ARGV[3]) else return 0 end`)
+	delkey    = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("DEL",KEYS[1]) else return 0 end`)
+	setkey    = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("SET",KEYS[1],ARGV[2],"PX",ARGV[3]) else return 0 end`)
+	setKeyLua = rueidis.NewLuaScript(`if redis.call("SET", KEYS[1], ARGV[1], "NX", "PX", ARGV[2]) then return KEYS[1] else return nil end`)
```
Comment (rueian) [inline rueidisaside/aside.go:214]: We should keep the script the same function as the SET NX GET.

File: rueidisaside/aside.go [hunk 3]
```diff
@@ -114,7 +117,7 @@ func (c *Client) keepalive() (id string, err error) {
 	c.mu.Unlock()
 	if id == "" {
 		id = PlaceholderPrefix + ulid.Make().String()
-		if err = c.client.Do(c.ctx, c.client.B().Set().Key(id).Value("").Px(c.ttl).Build()).Error(); err == nil {
+		if err = setKeyLua.Exec(c.ctx, c.client, []string{id}, []string{strconv.FormatInt(c.ttl.Milliseconds(), 10)}).Error(); err == nil {
```
Comment (rueian) [inline rueidisaside/aside.go:120]:
```suggestion
		if err = c.client.Do(c.ctx, c.client.B().Set().Key(id).Value("").Px(c.ttl).Build()).Error(); err == nil {
```

File: rueidisaside/aside.go [hunk 4]
```diff
@@ -197,6 +204,7 @@ func (c *Client) Close() {
 const PlaceholderPrefix = "rueidisid:"
 
 var (
-	delkey = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("DEL",KEYS[1]) else return 0 end`)
-	setkey = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("SET",KEYS[1],ARGV[2],"PX",ARGV[3]) else return 0 end`)
+	delkey    = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("DEL",KEYS[1]) else return 0 end`)
+	setkey    = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("SET",KEYS[1],ARGV[2],"PX",ARGV[3]) else return 0 end`)
+	setKeyLua = rueidis.NewLuaScript(`if redis.call("SET", KEYS[1], ARGV[1], "NX", "PX", ARGV[2]) then return KEYS[1] else return nil end`)
```
Comment (rueian) [inline rueidisaside/aside.go:209]: We should keep the script the same function as the SET NX GET. It should return nil when set and return the existing value when not set.

Comment (rueian) [reply rueidisaside/aside.go:209]: Hi @LegendHawk27, are you still interested in working on this?

Comment (LegendHawk27) [reply rueidisaside/aside.go:209]: Need some help to solve this issue

Comment (rueian) [reply rueidisaside/aside.go:209]: What help do you need?

File: rueidisaside/aside.go [hunk 5]
```diff
@@ -11,14 +11,15 @@ import (
 	"time"
 	"unsafe"
 
+	"github.com/oklog/ulid/v2"
```
Comment (rueian) [inline rueidisaside/aside.go:14]: This import is no longer needed.

File: rueidisaside/aside.go [hunk 6]
```diff
@@ -11,14 +11,15 @@ import (
...
 	"github.com/redis/rueidis"
 )
 
 type ClientOption struct {
-	// ClientBuilder can be used to modify rueidis.Client used by Locker
```
Comment (rueian) [inline rueidisaside/aside.go:18]: Please keep the comment.

File: rueidisaside/aside.go [hunk 7]
```diff
@@ -11,14 +11,15 @@ import (
...
 	ClientBuilder func(option rueidis.ClientOption) (rueidis.Client, error)
 	ClientOption  rueidis.ClientOption
-	ClientTTL     time.Duration // TTL for the client marker, refreshed every 1/2 TTL. Defaults to 10s. The marker allows other client to know if this client is still alive.
```
Comment (rueian) [inline rueidisaside/aside.go:21]: Please keep the comment.

File: rueidisaside/aside.go [hunk 8]
```diff
@@ -102,7 +105,7 @@ func (c *Client) refresh(id string) {
 			id2 := c.id
 			c.mu.Unlock()
 			if id2 != id {
-				return // client id has changed, abort this goroutine
```
Comment (rueian) [inline rueidisaside/aside.go:105]: Please keep the comment.

File: rueidisaside/aside.go [hunk 9]
```diff
@@ -116,7 +119,7 @@ func (c *Client) keepalive() (id string, err error) {
 	id = c.id
 	c.mu.Unlock()
 	if id == "" {
-		id = PlaceholderPrefix + randStr()
+		id = PlaceholderPrefix + ulid.Make().String()
```
Comment (rueian) [inline rueidisaside/aside.go:122]: We should use randStr() now.

File: rueidisaside/aside.go [hunk 10]
```diff
@@ -148,15 +151,20 @@ retry:
 	wait := c.register(key)
 	resp := c.client.DoCache(ctx, c.client.B().Get().Key(key).Cache(), ttl)
 	val, err := resp.ToString()
-	if rueidis.IsRedisNil(err) && fn != nil { // cache miss, prepare to populate the value by fn()
```
Comment (rueian) [inline rueidisaside/aside.go:151]: Please keep the comment.

File: rueidisaside/aside.go [hunk 11]
```diff
@@ -148,15 +151,20 @@ retry:
...
+				val, err = setKeyLua.Exec(ctx, c.client, []string{key}, []string{id, ttlStr}).ToString()
+			} else {
+				val, err = c.client.Do(ctx, c.client.B().Set().Key(key).Value(id).Nx().Get().Px(ttl).Build()).ToString()
+			}
+			if rueidis.IsRedisNil(err) {
 				if val, err = fn(ctx, key); err == nil {
-					err = setkey.Exec(ctx, c.client, []string{key}, []string{id, val, strconv.FormatInt(ttl.Milliseconds(), 10)}).Error()
+					err = setkey.Exec(ctx, c.client, []string{key}, []string{id, val, ttlStr}).Error()
 				}
-				if err != nil { // failed to populate the value, release the lock.
```
Comment (rueian) [inline rueidisaside/aside.go:159]: Please keep the comment.

File: rueidisaside/aside.go [hunk 12]
```diff
@@ -169,7 +177,6 @@ retry:
 		ph := c.register(val)
 		err = c.client.DoCache(ctx, c.client.B().Get().Key(val).Cache(), c.ttl).Error()
 		if rueidis.IsRedisNil(err) {
-			// the client who held the lock has gone, release the lock.
```
Comment (rueian) [inline rueidisaside/aside.go:172]: Please keep the comment.

File: rueidisaside/aside.go [hunk 13]
```diff
@@ -191,7 +198,6 @@ func (c *Client) Del(ctx context.Context, key string) error {
 	return c.client.Do(ctx, c.client.B().Del().Key(key).Build()).Error()
 }
 
-// Client exports the underlying rueidis.Client
```
Comment (rueian) [inline rueidisaside/aside.go:194]: Please keep the comment.

File: rueidisaside/aside.go [hunk 14]
```diff
@@ -210,6 +216,7 @@ func (c *Client) Close() {
 const PlaceholderPrefix = "rueidisid:"
 
 var (
-	delkey = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("DEL",KEYS[1]) else return 0 end`)
-	setkey = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("SET",KEYS[1],ARGV[2],"PX",ARGV[3]) else return 0 end`)
+	delkey    = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("DEL",KEYS[1]) else return 0 end`)
+	setkey    = rueidis.NewLuaScript(`if redis.call("GET",KEYS[1]) == ARGV[1] then return redis.call("SET",KEYS[1],ARGV[2],"PX",ARGV[3]) else return 0 end`)
+	setKeyLua = rueidis.NewLuaScript(`local set = redis.call("SET", KEYS[1], ARGV[1], "NX", "PX", ARGV[2]) if set then return nil else return redis.call("GET", KEYS[1]) end`)
```
Comment (rueian) [inline rueidisaside/aside.go:221]: The new script looks good to me. However, could we make it even shorter? for example:
```suggestion
	setKeyLua = rueidis.NewLuaScript(`if redis.call("SET", KEYS[1], ARGV[1], "NX", "PX", ARGV[2]) then return nil else return redis.call("GET", KEYS[1]) end`)
```

Besides that, we also need tests that have the `UseLuaScripts` enabled in the `aside_test.go` file. You can copy all the existing tests and set the `UseLuaScripts` for them.

**Pull request closed without merging**